### DISCUSSION TITLE
[Script request] Hoarder bookmarking/archiving app

### DISCUSSION THREAD

**Discussion by ptrsmk** [community-scripts/ProxmoxVE#73] [Request script]
### Application Name

Hoarder

### Website

https://hoarder.app

### Description

A self-hostable bookmark-everything app (links, notes and images) with AI-based automatic tagging and full text search. Apps available for Android and iOS, as well as plugins for Firefox and Chromium browsers.

I think this app has virtually everything one could want in an advanced bookmarking app, especially with the native support for Android and iOS. It's very actively developed and growing in popularity.

### Due Diligence

- [X] I have searched existing [scripts](https://community-scripts.github.io/Proxmox/scripts) and found no duplicates.
- [X] I have searched existing [discussions](https://github.com/community-scripts/ProxmoxVE/discussions?discussions_q=) and found no duplicate requests.
- [X] This is not a game-related request.

**Comment by ptrsmk** [comment]
There have been multiple requests for this already. 

https://github.com/tteck/Proxmox/discussions/3844
https://github.com/tteck/Proxmox/discussions/50#discussioncomment-11048928

**Comment by MickLesk** [comment]
I take a look tomorrow

**Reply by costa-alex** [reply]
any news? :)

**Reply by MickLesk** [reply]
yes, ive created a first prototype, but the documentation is bad, i fight with the script ^^

**Reply by MohamedBassem** [reply]
@MickLesk Hey, Hoarder's creator here. Sorry, you're struggling with the documentation. I'd be more than happy to help with any problems you're facing with it (and would love to hear your feedback to improve it). If you can join us on hoarder's [discord](https://discord.gg/NrgeYywsFh) that could be even faster.

**Reply by MickLesk** [reply]
> @MickLesk Hey, Hoarder's creator here. Sorry, you're struggling with the documentation. I'd be more than happy to help with any problems you're facing with it (and would love to hear your feedback to improve it). If you can join us on hoarder's [discord](https://discord.gg/NrgeYywsFh) that could be even faster.

Hey @MohamedBassem,

thanks for your response. Your documentation is of course only suitable for Docker Compose. We only offer bare metal installations, i.e. on a native e.g. Debian basis.
I've been struggling with the script for ages, either my development computer is too weak or it takes up too many resources. In the whole area of pnpm install and pnpm dlx the script flies away after some time (probably heap_space).

Hence the question in the first place:
1. what are the system requirements? You are not the only ones, all “modern” developers who only rely on Docker never talk about system requirements, neither CPU nor RAM is ever mentioned.

2. do you think it makes more sense to offer an Alpine installation only? Well, I have struggled with Debian and Ubunu for several hours, tried various things, but no result.
Here you can see the current install bash: https://raw.githubusercontent.com/MickLesk/Proxmox_DEV/refs/heads/main/install/hoarder-install.sh

You can reach me too in our discord: https://discord.gg/UHrpNWGwkH

**Reply by vhsdream** [reply]
Hi @MickLesk, just a few days ago I managed to install Hoarder bare-metal in a Debian 12 LXC. It currently works perfectly with 2GB RAM, 2 cores and 8GB of disk space to start. Crawling, OCR, search and integration with Ollama also work very well. I used the [AUR package](https://aur.archlinux.org/packages/hoarder)  as a guide since I'm familiar with the packaging system. It did have a couple of errors during the worker build phase, but they all seem to work.

Looking at the install script, I noticed a few things that I would say are essential, but forgive me if I am missing them. These are "optional" dependencies, but if you want it to work completely, I wouldn't consider the following optional:

- Headless Chrome/Chromium or Browserless: for web scraping; otherwise it just saves links.
- [Meilisearch](https://github.com/meilisearch/meilisearch) for searching through your collection. Search is disabled without it.
- yt-dl for downloading/archiving videos. I had to install a newer version from their Github; Debian's version did not work for me.
- [Monolith](https://github.com/Y2Z/monolith) for archiving webpages.

Lastly I see that postgres is installed, but Hoarder currently uses drizzle (some kind of ORM thing?) with sqlite and currently does not support postgres.

Limitations of the approach I took are that I just grabbed a .deb of Meilisearch and the Linux binary of Monolith and just threw them in the bin path, with little regard to any future issues. Also, following the AUR example results in 3 systemd service files - one for the workers, one for the web-server, and one for the browser. It's all a bit much and I'm not looking forward to updating it!

I'll see if I can find some time soon to fork your dev repo and mangle your script. I noticed you're working on an Immich script too (I've got a non-Docker install working in LXC the past few weeks with iGPU acceleration)!

**Reply by MickLesk** [reply]
Hi @vhsdream,
okay, thats fine, but how you build the scripts? With pnpm? That needes by my machine 12 GB RAM and 8 Cores at the moment.
I don't really think that much about the dependencies (Mellisearch), I'm interested in the PNPM builds for the worker, frontend and backend. :smile:

**Reply by vhsdream** [reply]
> Hi @vhsdream, okay, thats fine, but how you build the scripts? With pnpm? That needes by my machine 12 GB RAM and 8 Cores at the moment. I don't really think that much about the dependencies (Mellisearch), I'm interested in the PNPM builds for the worker, frontend and backend. 😄

Yes that's what I was saying - I followed nearly the same build process I see in your dev script, and I did it inside an LXC with 2GB RAM and 2 cores. 

I forked the dev repo and will make some changes to add the dependences and try to build it again, first using the method in your script, then with the way I did it (the only difference I can see right now is I ran `corepack enable` beforehand, which ends up downloading a specific version of pnpm to use).

**Reply by MickLesk** [reply]
If you have some time, can you generate a script / or just a blunt command sequence so that I can recreate it? Maybe I'm now operating blind with Hoarder 😃

**Reply by MohamedBassem** [reply]
Ok, so I spent some time debugging this. Something weird is happening during the installation on debian indeed. It shouldn't take that long, and I'm still yet to figure out what the probelm is.`postman-code-generators` postinstall scripts are taking forever for some reason, and this doesn't happen on alpine for example.

**Reply by MohamedBassem** [reply]
Ok, I managed to figure out the reason why the installation was getting stuck. I've sent (https://github.com/MickLesk/Proxmox_DEV/pull/3) that explains on a high level how the script should look like. I noticed that @vhsdream also sent a PR which is more comprehensive than mine. I think final script will be a mix of both PRs.

**Reply by MickLesk** [reply]
Oh Fckn nice. I didn't even see that because I have the repo on ignore :smile:

Thank you both so much, I'll have a look at it straight away tomorrow morning! 👍😊

**Reply by vhsdream** [reply]
I've updated my PR in your dev repo, it should be good to go aside from the update function 👍🏻

**Reply by vhsdream** [reply]
@MickLesk Now I've done a PR for the update_script function - check it out when you have some time 😄

**Reply by MickLesk** [reply]
Script runs fine, i need to finetune some things, but it works smooth 😄 Thx for help!

The PR needs a little bit time, because ive develop big changes in the core (if you see when you run the bash) 😄 

```bash
bash -c "$(wget -qLO - https://raw.githubusercontent.com/MickLesk/Proxmox_DEV/refs/heads/main/ct/hoarder.sh)"
```

**Comment by vhsdream** [comment]
@MickLesk - I have adapted our mutual work on the install and update script for Hoarder and have a working example. I wanted to know if it is OK if I submitted a PR in this repo?

**Reply by MickLesk** [reply]
Yep

**Reply by MickLesk** [reply]
before you prepare an PR, one $STD is missing:

<img width="344" alt="image" src="https://github.com/user-attachments/assets/ffad6348-17fd-4323-ae8e-5dc14af63e71">

**Comment by vhsdream** [comment]
PR submitted - #567
